doc/design/persistent_volumes_etcd_data.md

```diff
@@ -0,0 +1,79 @@
+
+When a majority of pods is deleted at the same time the etcd cluster will become unquorate and there'll be no way to restore its quorum since the members and their data are gone. So the unique solution (that etcd-operator currently do) is to reinitialize a new cluster restoring its data from a backup (if available).
+
+This proposal is to fix such case achieving a persistent/durable etcd cluster.
+
+## Proposed Changes
+
+#### Changes
+
+Add an option to enable storing etcd data inside a persistent volume and related reconcile logic changes.
+
+The change to be effective and cover the above failure scenario is split in two parts.
+
+### Part 1
+
+- Add to the cluster spec PodPolicy an option to enable putting etcd data inside a Persistent Volume
+- If podPolicy has persistent volumes enabled, when creating a pod also create a Persistent Volume Claim and use it as the volume source for etcd-data
+- Handle etcd data persistent volume garbage collection like already done with other resources. While backup pv aren't garbage collected, etcd data pv needs to be garbage collected when not needed anymore.
+
+#### Pod name to Persistent Volume Claim name mapping
+
+The pod and related PVC are mapped 1 to 1 so thet can be easily generated and deducted. I.E. if the pod is called `etcd-0000` its related PVC name will be `etcd-0000-pvc`
+
+#### Results
+
+These changes will put the basis for initial persistent volume management BUT won't fix the failure scenario described at the start of the proposal: if a majority of pods is deleted there's no logic to recreate them with their previous data.
+
+#### Proposed enhancements to Part 1 (@xiang90)
+
+- When using a PVC as volume source also set pod restart policy to `Always`
+
+With this change when the etcd container of the pod dies due to different reasons (etcd crashing, container killed, node reboots but its pods aren't evicted by the k8s node controller) it will be restarted (since Restart Policy is set to Always). Previously the pod will go in Failed state and will be removed and etcd operator will remove the etcd member and add a new member and a new pod.
+
+
+### Part 2
+
+#### Changes
+
+- Change reconcile logic to handle deleted pods that had a persistent volume:
+
+If the a pod of an etcd cluster member is deleted and it had a persistent volume claim associated, instead of removing the etcd member from the etcd cluster and adding a new member just recreate it (keeping its pod name) binding the existing persistent volume claim. The pod recreation will happen only if the pod doesn't exists anymore in the API because we'll keep the same pod name. This also enable a MANDATORY property: at most once pod existence logic to avoid having in the API at the same time two pods with associated the same PVC.
```

Comment:
> If the a pod of an etcd cluster member is deleted

Please list what scenarios a member will be deleted. For this phase it will be deleted only in scale-down

Reply:
> instead of removing the etcd member from the etcd cluster and adding a new member just recreate it (keeping its pod name) binding the existing persistent volume claim.

For this phase, we assume that we will wait for any failed etcd pod to restart instead of recreating it.

Handling permanent failure is a more complicated topic. Like how to manage membership and how to define failure/tolerations.

Reply:
> For this phase, we assume that we will wait for any failed etcd pod to restart instead of recreating it.

This is handled by part one, but as said pods can be removed from the API and Part 2 aims to recreate them.



> Handling permanent failure is a more complicated topic. Like how to manage membership and how to define failure/tolerations.

part 2 isn't for handling permament failures but pods removed from the API.

Reply:
> This is handled by part one, but as said pods can be removed from the API and Part 2 aims to recreate them.

This is mentioned in [this discussion](https://github.com/coreos/etcd-operator/pull/1374/files#r136131112). We will solve them one by one, not coupled together.

> part 2 isn't for handling permanent failures but pods removed from the API.

What is the problem you are trying to solve? Can you give specific failure scenario and differentiation?

Reply:
I added at the start of the proposal 4 example cases used to better explain what I want to solve and the effect of every change on these cases.